crates/bevy_ecs/src/storage/sorted_small_vec.rs

```diff
@@ -0,0 +1,259 @@
+        Intersection {
+            this: self,
+            other,
+            i: 0,
+            j: 0,
+        }
+    }
+
+    /// Return the elements that are in `self` but not in `other`.
+    pub(crate) fn difference<'a>(&'a self, other: &'a Self) -> Difference<'a, N> {
+        Difference {
+            this: self,
+            other,
+            i: 0,
+            j: 0,
+        }
+    }
+
+    /// Returns true if the two vectors have no common elements.
+    pub(crate) fn is_disjoint(&self, other: &Self) -> bool {
+        self.intersection(other).next().is_none()
+    }
+
+    /// Returns true if all the elements in `self` are also in `other`.
+    pub(crate) fn is_subset(&self, other: &Self) -> bool {
+        self.difference(other).next().is_none()
+    }
+}
+
+impl<const N: usize> Extend<usize> for SortedSmallVec<N> {
+    fn extend<T: IntoIterator<Item = usize>>(&mut self, other: T) {
+        let mut i = 0;
+        let mut other_iter = other.into_iter();
+        let mut other_val = other_iter.next();
+        while i < self.len() && other_val.is_some() {
+            let val_j = other_val.unwrap();
+            match self.0[i].cmp(&val_j) {
+                Ordering::Less => {
+                    i += 1;
+                }
+                Ordering::Greater => {
+                    self.0.insert(i, val_j);
+                    other_val = other_iter.next();
+                }
+                Ordering::Equal => {
+                    i += 1;
+                    other_val = other_iter.next();
+                }
+            }
+        }
+        while let Some(val_j) = other_val {
+            self.0.push(val_j);
+            other_val = other_iter.next();
+        }
+    }
+}
+
+/// Intersection between `this` and `other` sorted vectors.
+pub(crate) struct Intersection<'a, const N: usize> {
+    this: &'a SortedSmallVec<N>,
+    other: &'a SortedSmallVec<N>,
+    i: usize,
+    j: usize,
+}
+
+impl<'a, const N: usize> Iterator for Intersection<'a, N> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut res = None;
+        while self.i < self.this.len() && self.j < self.other.len() {
+            if (self.i == 0 || self.this.0[self.i] != self.this.0[self.i - 1])
+                && self.this.0[self.i] == self.other.0[self.j]
+            {
+                res = Some(self.this.0[self.i]);
+                self.i += 1;
+                self.j += 1;
+                return res;
+            } else if self.this.0[self.i] < self.other.0[self.j] {
+                self.i += 1;
+            } else {
+                self.j += 1;
+            }
+        }
+        res
+    }
+}
+
+impl<'a, const N: usize> From<Intersection<'a, N>> for SortedSmallVec<N> {
+    fn from(intersection: Intersection<'a, N>) -> Self {
+        let mut sorted_vec = SortedSmallVec::new_const();
+        for value in intersection.into_iter() {
```

Comment:
I think we can use a cheaper impl here. Since we know Intersection returns elements in sorted order, we can just `SmallVec::from_iter(intersection)` I think.